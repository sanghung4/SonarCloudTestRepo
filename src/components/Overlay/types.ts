import { GestureResponderEvent } from 'react-native';

export interface NeedHelpProps {
  onDismissRequest?: (event: GestureResponderEvent) => void;
}

export enum OVERLAY_TEST_IDS {
  ALERT = 'alertTest',
  OVERLAY = 'overlayTest',
}
