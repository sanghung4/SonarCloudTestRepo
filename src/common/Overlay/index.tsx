import { ReactNode } from 'react';

import { Alert, Box, Grid } from '@dialexa/reece-component-library';

type Props = {
  show?: boolean;
  infoMessages?: string[];
  children: ReactNode;
};

function Overlay(props: Props) {
  return (
    <Box position="relative">
      <Box
        position="absolute"
        width={1}
        height={1}
        bgcolor={props.show ? 'common.white' : 'transparent'}
        sx={{ opacity: props.show ? 0.6 : 0 }}
      />
      {!!props.show && (
        <Box position="absolute" width={1} height={1} zIndex={999}>
          <Grid container alignItems="center" justifyContent="center">
            <Grid item md={6} xs={12}>
              <Alert
                severity="info"
                variant="filled"
                sx={{ color: 'common.white' }}
              >
                {props.infoMessages?.map((m: string) => (
                  <Box key={m}>{m} </Box>
                ))}
              </Alert>
            </Grid>
          </Grid>
        </Box>
      )}
      {props.children}
    </Box>
  );
}

export default Overlay;
