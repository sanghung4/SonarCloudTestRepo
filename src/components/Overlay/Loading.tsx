import { Colors } from 'constants/style';
import React from 'react';
import { ActivityIndicator } from 'react-native';

const Loading = () => {
  return <ActivityIndicator size="large" color={Colors.PRIMARY_2100} />;
};

export default Loading;
