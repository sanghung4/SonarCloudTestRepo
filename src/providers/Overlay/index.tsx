import React, { FC, createContext, useState, useContext } from 'react';
import { SortByEnum } from 'screens/VarianceLocationList/types';
import {
  OverlayContextValues,
  AlertPayload,
  AlertAction,
  OverlayTypes,
} from './types';

const initialAlertAction: AlertAction = {
  title: 'Ok',
  onPress: (): void => {},
};

const initialAlertPayload: AlertPayload = {
  title: '',
  description: '',
  actions: [],
  options: {},
};

const initialContext: OverlayContextValues = {
  alertPayload: initialAlertPayload,
  visibleOverlay: '',
  showAlert: () => {},
  hideAlert: () => {},
  toggleLoading: () => {},
  toggleSortBy: () => {},
  handleSortBy: () => {},
  activeSortBy: SortByEnum.ASCENDING,
};

const OverlayContext = createContext(initialContext);

const OverlayProvider: FC = ({ children }) => {
  const [alertPayload, setAlertPayload] = useState(initialAlertPayload);
  const [visibleOverlay, setVisibleOverlay] = useState<OverlayTypes>('');
  const [activeSortBy, setActiveSortBy] = useState<SortByEnum>(
    SortByEnum.ASCENDING
  );

  const showAlert = (payload: AlertPayload) => {
    setAlertPayload({
      ...initialAlertPayload,
      actions: [initialAlertAction],
      ...payload,
    });
    setVisibleOverlay('alert');
  };

  const hideAlert = () => {
    setAlertPayload(initialAlertPayload);
    setVisibleOverlay('');
  };

  const toggleLoading = (state: boolean) => {
    setVisibleOverlay(state ? 'loading' : '');
  };

  const handleSortBy = (state: SortByEnum) => {
    setActiveSortBy(state);
  };

  const toggleSortBy = (state: boolean) => {
    setVisibleOverlay(state ? 'sortBy' : '');
  };

  return (
    <OverlayContext.Provider
      value={{
        alertPayload,
        visibleOverlay,
        showAlert,
        hideAlert,
        toggleLoading,
        toggleSortBy,
        handleSortBy,
        activeSortBy,
      }}
    >
      {children}
    </OverlayContext.Provider>
  );
};

export default OverlayProvider;

export const useOverlay = (): OverlayContextValues =>
  useContext(OverlayContext);
