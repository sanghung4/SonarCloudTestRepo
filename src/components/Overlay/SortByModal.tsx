import React, { useState, useEffect } from 'react';
import { StyleSheet, TouchableOpacity, View } from 'react-native';
import { Text } from 'components/Text';
import { Colors, fontSize, FontWeight } from 'constants/style';
import { useOverlay } from 'providers/Overlay';
import { SortByEnum } from 'screens/VarianceLocationList/types';
import { CustomButton } from 'components/CustomButton';
import { CustomIcon, CustomIconNames } from 'components/CustomIcon';

const SortByModal = () => {
  const { toggleSortBy, handleSortBy, activeSortBy } = useOverlay();

  const [tempSortBy, setActiveSortBy] = useState<SortByEnum>(
    SortByEnum.ASCENDING
  );

  const options = [
    SortByEnum.ASCENDING,
    SortByEnum.DESCENDING,
    SortByEnum.NET_HIGHTOLOW,
    SortByEnum.NET_LOWTOHIGH,
    SortByEnum.GROSS_HIGHTOLOW,
    SortByEnum.GROSS_LOWTOHIGH,
  ];

  useEffect(() => {
    if (tempSortBy !== activeSortBy) {
      setActiveSortBy(activeSortBy);
    }
  }, []);

  return (
    <View style={styles.sortByContainer}>
      <CustomIcon
        name={CustomIconNames.Delete}
        style={styles.deleteIcon}
        onPress={() => toggleSortBy(false)}
      />
      <Text style={styles.sortByText}>Sort By</Text>
      {options.map((option: SortByEnum) => {
        return (
          <TouchableOpacity
            key={option}
            style={styles.optionContainer}
            onPress={() => {
              setActiveSortBy(option);
            }}
          >
            <CustomIcon
              name={
                option === tempSortBy
                  ? CustomIconNames.RadioActive
                  : CustomIconNames.RadioInactive
              }
            />
            <Text style={styles.optionText}>{option}</Text>
          </TouchableOpacity>
        );
      })}
      <View style={styles.actionContainer}>
        <CustomButton
          title={'Cancel'}
          onPress={() => toggleSortBy(false)}
          type={'link'}
        />
        <CustomButton
          title={'Sort'}
          buttonStyle={styles.sortButton}
          onPress={() => {
            handleSortBy(tempSortBy);
            toggleSortBy(false);
          }}
        />
      </View>
    </View>
  );
};
const styles = StyleSheet.create({
  actionContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-evenly',
    width: '100%',
  },
  cancelButton: {
    width: 137,
    height: 44,
    justifyContent: 'center',
    alignItems: 'center',
  },
  cancelButtonText: {
    fontSize: fontSize.BASE,
    fontWeight: FontWeight.MEDIUM,
    color: Colors.PRIMARY_3100,
    textDecorationLine: 'underline',
  },
  deleteIcon: {
    alignSelf: 'flex-end',
  },
  optionContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    marginLeft: 23,
    marginVertical: 14,
  },
  optionText: {
    marginLeft: 23,
  },
  sortButton: {
    width: 137,
    height: 44,
  },
  sortButtonText: {
    fontSize: fontSize.BASE,
    fontWeight: FontWeight.MEDIUM,
    color: Colors.WHITE,
  },
  sortByContainer: {
    width: 325,
    minHeight: 331,
    backgroundColor: Colors.WHITE,
    padding: 12,
  },
  sortByText: {
    marginLeft: 21,
    fontWeight: FontWeight.MEDIUM,
    color: Colors.PRIMARY_1100,
    fontSize: fontSize.H5,
  },
});

export default SortByModal;
