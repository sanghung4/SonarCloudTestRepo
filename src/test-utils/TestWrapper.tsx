import React, { JSXElementConstructor, ReactElement, ReactNode } from 'react';

import { ThemeProvider } from '@dialexa/reece-component-library';
import { MockedProvider, MockedResponse } from '@apollo/client/testing';
import { OktaAuth } from '@okta/okta-auth-js';
import { render as jestRender, RenderResult } from '@testing-library/react';
import { I18nextProvider } from 'react-i18next';
import { MemoryRouter } from 'react-router-dom';

import { UserContextType, AuthContext } from 'AuthProvider';
import { Permission } from 'common/PermissionRequired';
import i18n from 'test-utils/mockI18N';
import {
  SelectedAccountsContext,
  SelectedAccountsContextType
} from 'providers/SelectedAccountsProvider';
import { CartContext, CartContextType } from 'providers/CartProvider';
import {
  CheckoutContext,
  CheckoutContextType,
  initialCheckoutContextValues
} from 'Checkout/CheckoutProvider';

/**
 * NOTE:
 * To setup mocks follow the examples here:
 * https://www.apollographql.com/docs/react/development-testing/testing/
 *
 * To get a successful response in your test (not just the loading state)
 * you can do the following:
 *
 * it('Should render', async () => {
 *   const { container } = renderMockRouter(<MyComponent />, mocks);
 *
 *   await waitFor(() => new Promise((res) => setTimeout(res, 0)));
 *
 *   expect(container).toMatchSnapshot();
 * });
 *
 */

/******************************/
/* Types                      */
/******************************/
type Props = {
  children?: ReactNode;
  authConfig?: Partial<UserContextType>;
  selectedAccountsConfig?: Partial<SelectedAccountsContextType>;
  cartConfig?: Partial<CartContextType>;
  checkoutConfig?: Partial<CheckoutContextType>;
  mocks?: ReadonlyArray<MockedResponse>;
};

type WrapperProps<TKey extends keyof Props> = {
  value: Props[TKey];
  children: React.ReactNode;
};

/******************************/
/* Defaults                   */
/******************************/
const defaultAuth: UserContextType = {
  authState: {
    isAuthenticated: true
  },
  isLoggingOut: false,
  oktaAuth: new OktaAuth({ issuer: 'http://test.com' }),
  profile: {
    userId: 'testuser',
    permissions: Object.values(Permission),
    isEmployee: false,
    isVerified: false
  }
};

const defaultSelectedAccounts: SelectedAccountsContextType = {
  clearAccounts: () => {},
  updateAccounts: () => {},
  updateShippingBranchId: () => {},
  loading: false,
  selectedAccounts: {
    billTo: { erpAccountId: '' },
    shipTo: { erpAccountId: '' },
    shippingBranchId: ''
  },
  isMincron: false,
  isEclipse: false
};

const defaultCartContextValue: CartContextType = {
  addItemToCart: () => {},
  addItemsToCart: () => {},
  addAllListItemsToCart: () => {},
  cart: undefined,
  cartLoading: false,
  checkingOutWithQuote: false,
  contractBranch: undefined,
  clearContract: () => {},
  clearQuote: () => {},
  deleteCartItems: () => new Promise(() => {}),
  disableAddToCart: false,
  deleteItem: () => {},
  getUserCart: () => new Promise(() => {}),
  itemCount: 0,
  itemLoading: undefined,
  lineNotes: {},
  releaseContractToCart: () => {},
  setContract: () => {},
  setLineNotes: () => {},
  setQuoteId: () => {},
  setQuoteData: () => {},
  setQuoteShipToId: () => {},
  setPreviousCart: () => {},
  setSelectedBranch: () => {},
  updateCart: undefined,
  updateDelivery: undefined,
  updateItemQuantity: () => {},
  updateWillCall: undefined,
  updateWillCallBranch: undefined,
  refreshCart: undefined,
  isWillCall: false,
  deleteCreditCardFromCart: () => new Promise(() => {})
};

/******************************/
/* Wrappers                   */
/******************************/
function SelectedAccountsProviderWrapper({
  value,
  children
}: WrapperProps<'selectedAccountsConfig'>) {
  if (value) {
    return (
      <SelectedAccountsContext.Provider
        value={{ ...defaultSelectedAccounts, ...value }}
      >
        {children}
      </SelectedAccountsContext.Provider>
    );
  }

  return <>{children}</>;
}

function CartProviderWrapper({ value, children }: WrapperProps<'cartConfig'>) {
  if (value) {
    return (
      <CartContext.Provider value={{ ...defaultCartContextValue, ...value }}>
        {children}
      </CartContext.Provider>
    );
  }

  return <>{children}</>;
}

function CheckoutProviderWrapper({
  value,
  children
}: WrapperProps<'checkoutConfig'>) {
  if (value) {
    return (
      <CheckoutContext.Provider
        value={{ ...initialCheckoutContextValues, ...value }}
      >
        {children}
      </CheckoutContext.Provider>
    );
  }

  return <>{children}</>;
}

function TestWrapper(props: Props) {
  return (
    <ThemeProvider>
      <I18nextProvider i18n={i18n}>
        <AuthContext.Provider value={{ ...defaultAuth, ...props.authConfig }}>
          <MockedProvider mocks={props.mocks ?? []}>
            <SelectedAccountsProviderWrapper
              value={props.selectedAccountsConfig}
            >
              <CartProviderWrapper value={props.cartConfig}>
                <CheckoutProviderWrapper value={props.checkoutConfig}>
                  {props.children}
                </CheckoutProviderWrapper>
              </CartProviderWrapper>
            </SelectedAccountsProviderWrapper>
          </MockedProvider>
        </AuthContext.Provider>
      </I18nextProvider>
    </ThemeProvider>
  );
}

function jestRenderFn(wrapped: JSX.Element) {
  return jestRender(wrapped, {
    wrapper: MemoryRouter as
      | JSXElementConstructor<{
          children: ReactElement<any, string | JSXElementConstructor<any>>;
        }>
      | undefined
  });
}

export function render(component: ReactNode, config?: Props) {
  const wrapped = <TestWrapper {...config}>{component}</TestWrapper>;
  return jestRenderFn(wrapped);
}
export type TestRender = RenderResult<
  typeof import('@testing-library/dom/types/queries'),
  Element | DocumentFragment
>;

export function renderWithComponent(component: ReactNode, config?: Props) {
  const wrapped = <TestWrapper {...config}>{component}</TestWrapper>;
  return {
    render: jestRenderFn(wrapped),
    component: wrapped
  };
}

export default TestWrapper;
