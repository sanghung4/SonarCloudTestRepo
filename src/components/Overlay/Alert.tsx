import React from 'react';
import { Dimensions, View } from 'react-native';
import { Text } from 'components/Text';
import * as SVG from 'components/SVG';
import { AlertOptions } from 'providers/Overlay/types';
import { useOverlay } from 'providers/Overlay';
import { Colors, FontWeight } from 'constants/style';
import { CustomIcon, CustomIconNames } from 'components/CustomIcon';
import { CustomButton } from 'components/CustomButton';
import { getScreenTestingIds } from 'test-utils/testIds';

const Alert = () => {
  const {
    alertPayload: {
      svg,
      title,
      description,
      error,
      actions,
      onClose,
      options = {
        fullScreen: false,
        centered: false,
        hideCloseButton: false,
      },
    },
    hideAlert,
  } = useOverlay();

  const handleClose = () => {
    hideAlert();
    onClose && onClose();
  };

  const testIds = getScreenTestingIds('Alert');

  const SVGComponent = svg ? SVG[svg] : (_: SVG.SVGProps) => null;

  return (
    <View
      style={[
        styles.alertContainer,
        options.fullScreen ? styles.alertFullScreen : styles.alertNotFullScreen,
      ]}
      testID={testIds.component}
    >
      <SVGComponent size={180} style={styles.svg(options)} />

      {/* Close Button */}
      {!options.fullScreen && (
        <CustomIcon
          name={CustomIconNames.Delete}
          onPress={handleClose}
          containerStyle={styles.closeButton}
          testID={testIds.closeIcon}
        />
      )}

      {/* Title */}
      <Text
        fontSize={20}
        fontWeight={FontWeight.MEDIUM}
        color={Colors.PRIMARY_1100}
        centered={options.centered}
        testID={testIds.title}
      >
        {title}
      </Text>

      {/* Description */}
      <Text
        fontSize={16}
        fontWeight={FontWeight.REGULAR}
        color={Colors.SECONDARY_2100}
        centered={options.centered}
        style={[styles.textContent, !options.fullScreen && styles.flexGrow]}
        testID={testIds.description}
      >
        {description}
      </Text>

      {/* Error */}
      <Text
        caption
        color={Colors.SUPPORT_2100}
        centered={options.centered}
        style={styles.textContent}
        testID={testIds.error}
      >
        {error}
      </Text>

      {/* Actions */}
      <View style={styles.actions}>
        {Array.isArray(actions)
          ? actions.map(({ onPress, ...action }, index) => (
              <CustomButton
                key={index}
                onPress={() => {
                  onPress();
                  hideAlert();
                }}
                {...action}
                testID={`${testIds.action}-${action.key}`}
              />
            ))
          : null}
      </View>
    </View>
  );
};
const { width } = Dimensions.get('screen');

const styles = {
  svg: ({ centered }: AlertOptions = {}) => ({
    marginBottom: 18,
    ...(centered ? ({ alignSelf: 'center' } as const) : {}),
  }),

  alertContainer: {
    padding: 32,
  },
  alertFullScreen: {
    alignItems: 'center',
    justifyContent: 'center',
    flex: 1,
  },
  alertNotFullScreen: {
    width: width - 48,
    maxWidth: 360,
  },
  textContent: {
    marginTop: 12,
  },
  flexGrow: {
    flexGrow: 1,
  },

  actions: {
    marginTop: 20,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
  },
  actionContainer: {
    flex: 1,
  },
  action: {
    padding: 8,
  },
  closeButton: { position: 'absolute', right: 12, top: 12 },
} as const;

export default Alert;
