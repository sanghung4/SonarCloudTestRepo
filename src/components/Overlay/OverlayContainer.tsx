import React from 'react';
import { StyleSheet } from 'react-native';
import { Overlay } from 'react-native-elements';
import { useOverlay } from 'providers/Overlay';

import Alert from './Alert';
import Loading from './Loading';
import SortByModal from './SortByModal';
import { getScreenTestingIds } from 'test-utils/testIds';

const OverlayContainer: React.FC = () => {
  const { alertPayload, visibleOverlay } = useOverlay();

  const testIds = getScreenTestingIds('Overlay');

  const getTestId = () => {
    if (visibleOverlay === 'alert') {
      return testIds.alert;
    }
    if (visibleOverlay === 'loading') {
      return testIds.loading;
    }
    if (visibleOverlay === 'sortBy') {
      return testIds.sortyBy;
    }
  };

  return (
    <Overlay
      testID={getTestId()}
      isVisible={
        visibleOverlay === 'alert' ||
        visibleOverlay === 'loading' ||
        visibleOverlay === 'sortBy'
      }
      fullScreen={
        !!alertPayload.options?.fullScreen || visibleOverlay === 'loading'
      }
      overlayStyle={[
        styles.overlay,
        visibleOverlay === 'loading' && styles.loadingOverlay,
      ]}
    >
      {visibleOverlay === 'alert' && <Alert />}
      {visibleOverlay === 'loading' && <Loading />}
      {visibleOverlay === 'sortBy' && <SortByModal />}
    </Overlay>
  );
};

const styles = StyleSheet.create({
  overlay: { padding: 0, borderRadius: 2 },
  loadingOverlay: {
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: 'transparent',
  },
});

export default OverlayContainer;
