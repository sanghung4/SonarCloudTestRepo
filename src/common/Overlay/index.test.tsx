import { render } from 'test-utils/TestWrapper';
import Overlay from 'common/Overlay';

const mockChildren = <div>test</div>;
describe('common - Overlay', () => {
  it('Should match snapshot with the bare minimum props', () => {
    const { container } = render(<Overlay>{mockChildren}</Overlay>);
    expect(container).toMatchSnapshot();
  });

  it('Should match snapshot with the valid props', () => {
    const msg = ['test1', 'test2', 'test3'];
    const { container } = render(
      <Overlay show infoMessages={msg}>
        {mockChildren}
      </Overlay>
    );
    expect(container).toMatchSnapshot();
  });
});
