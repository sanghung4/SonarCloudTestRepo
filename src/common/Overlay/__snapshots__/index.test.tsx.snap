// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`common - Overlay Should match snapshot with the bare minimum props 1`] = `
<div>
  <div
    class="MuiBox-root css-79elbk"
  >
    <div
      class="MuiBox-root css-yedgod"
    />
    <div>
      test
    </div>
  </div>
</div>
`;

exports[`common - Overlay Should match snapshot with the valid props 1`] = `
<div>
  <div
    class="MuiBox-root css-79elbk"
  >
    <div
      class="MuiBox-root css-1sd540e"
    />
    <div
      class="MuiBox-root css-1lusoco"
    >
      <div
        class="MuiGrid-root MuiGrid-container css-go2vvz-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6 css-1m6pi2w-MuiGrid-root"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAlert-root MuiAlert-filledInfo MuiAlert-filled css-mrp2hz-MuiPaper-root-MuiAlert-root"
            role="alert"
          >
            <div
              class="MuiAlert-icon css-1ytlwq5-MuiAlert-icon"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="InfoIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z"
                />
              </svg>
            </div>
            <div
              class="MuiAlert-message css-acap47-MuiAlert-message"
            >
              <div
                class="MuiBox-root css-0"
              >
                test1
                 
              </div>
              <div
                class="MuiBox-root css-0"
              >
                test2
                 
              </div>
              <div
                class="MuiBox-root css-0"
              >
                test3
                 
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div>
      test
    </div>
  </div>
</div>
`;
